### PR TITLE
made us to clean up all live errors for projects that built cleanly

### PR DESCRIPTION
**Pull request by heejaechang**
it turns out this ability already existed as option. some tweak to make it behave a bit better. so we will clean up all errors in a project if that project built cleanly by default.

there is still an option to do so even for project that failed to build due to errors.

### CODE REVIEW

**Comment by heejaechang** [comment]
@dotnet/roslyn-analysis can you take a look?

**Comment by heejaechang** [comment]
@mavasani @srivatsn ping?

**Comment by heejaechang** [comment]
ping?

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs [hunk 1]
```diff
@@ -20,6 +21,12 @@ internal partial class DiagnosticIncrementalAnalyzer
         /// </summary>
         private partial class StateManager
         {
+            // this cache is supposed to be most of time empty except when we are doing build-live deduplication.
+            // since build is not something that happens continuously, we do not want to hold this onto memory all the time. we will build it once
+            // per build even if it means we are building same one as last time. this cache will go away as soon as solution snapshot moves which
+            // is quite frequent since it includes thing like open/close document, typing, file change and etc.
+            private static readonly ConditionalWeakTable<Project, List<StateSet>> s_buildOnlyProjectStateSetCache = new ConditionalWeakTable<Project, List<StateSet>>();
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs]: We might need to do performance measurements - last time I used CWT from Project to CompilationWithAnalyzers, it caused VM to rise quite a bit.

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs [hunk 2]
```diff
@@ -116,6 +132,38 @@ public void RemoveStateSet(ProjectId projectId)
                 _projectStates.RemoveStateSet(projectId);
             }
 
+            private List<StateSet> CreateBuildOnlyProjectStateSet(Project project)
+            {
+                var referenceIdentities = project.AnalyzerReferences.Select(r => _analyzerManager.GetAnalyzerReferenceIdentity(r)).ToSet();
+                var stateSetMap = GetStateSets(project).ToDictionary(s => s.Analyzer, s => s);
+
+                var stateSets = new List<StateSet>();
+                var analyzerMap = _analyzerManager.GetHostDiagnosticAnalyzersPerReference(project.Language);
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs]: Why are we using host analyzer state sets for build only diagnostics?

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs]: I see, then, I will make it explicit.

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs (hunk 2 above)
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs]: we are not, this code is to remove stateset that are only for host analyzers. this code is to see whether there is analyzer that exist both in project and host, and if they do, we include it here.

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs [hunk 3]
```diff
@@ -116,6 +132,38 @@ public void RemoveStateSet(ProjectId projectId)
...
+                    if (!referenceIdentities.Contains(identity))
+                    {
+                        // it is from host analyzer package rather than project analyzer reference
+                        // which build doesn't have
+                        continue;
+                    }
+
+                    // if same analyzer exists both in host (vsix) and in analyzer reference,
+                    // we include it in build only analyzer.
+                    foreach (var analyzer in kv.Value)
```
Comment (heejaechang) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs]: this is comment that explains it.

File: src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs (hunk 2 above)
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.StateManager.cs]: Ah, thanks. Can you please add a comment on the same lines?

**Comment by mavasani** [comment]
:+1: except the CWT part - which may need perf measurements.

**Comment by heejaechang** [comment]
@mavasani made cache explicit.

**Pull request merged**